Address Safer CPP warnings in WKWebView.mm

cdumez opened the pull request:
#### d054486edb1875e29958ad784efaaa000b49c167
<pre>
Address Safer CPP warnings in WKWebView.mm
<a href="https://bugs.webkit.org/show_bug.cgi?id=299688">https://bugs.webkit.org/show_bug.cgi?id=299688</a>

Reviewed by Darin Adler.

* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(nsErrorFromExceptionDetails):
(-[WKWebView _initializeWithConfiguration:]):
(-[WKWebView _didInsertAttachment:withSource:]):
(-[WKWebView _didRemoveAttachment:]):
(-[WKWebView _didInvalidateDataForAttachment:]):
(-[WKWebView createWebArchiveDataWithCompletionHandler:]):
(dictionaryRepresentationForEditorState):
(-[WKWebView _didChangeEditorState]):
(-[WKWebView setUnderPageBackgroundColor:]):
(-[WKWebView _updateFixedContainerEdges:]):
(createUserInfo):
(-[WKWebView _dataTaskWithRequest:runAtForegroundPriority:completionHandler:]):
(-[WKWebView _loadAndDecodeImage:constrainedToSize:maximumBytesFromNetwork:completionHandler:]):
(-[WKWebView _decodeImageData:preferredSize:completionHandler:]):
(-[WKWebView _getMainResourceDataWithCompletionHandler:]):
(-[WKWebView _createWebArchiveForFrame:completionHandler:]):
(-[WKWebView _createWebArchiveForFrames:rootFrame:completionHandler:]):
(-[WKWebView _getApplicationManifestWithCompletionHandler:]):
(-[WKWebView _getTextFragmentMatchWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

Canonical link: <a href="https://commits.webkit.org/300671@main">https://commits.webkit.org/300671@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d2b88509a2b44077e58003e9a7ac378115c30d30

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123447 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43162 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33858 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130165 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75583 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a2642e52-e238-4cd3-805e-500c78c92cba) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43885 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51756 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93835 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62280 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6caff11d-6fca-464b-b1d6-8f92c6ab1ab0) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126400 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34966 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110444 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74466 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/12d828c2-27ad-4fa4-aee3-d93d8fab5ec9) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33935 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73682 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104681 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28827 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132884 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50397 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38360 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102328 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50773 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106666 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102180 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25973 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47535 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25764 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47202 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50252 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56013 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49724 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53073 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51401 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -306,8 +306,8 @@
     [userInfo setObject:details->message.createNSString().get() forKey:_WKJavaScriptExceptionMessageErrorKey];
-    [userInfo setObject:@(details->lineNumber) forKey:_WKJavaScriptExceptionLineNumberErrorKey];
-    [userInfo setObject:@(details->columnNumber) forKey:_WKJavaScriptExceptionColumnNumberErrorKey];
+    [userInfo setObject:RetainPtr { @(details->lineNumber) }.get() forKey:_WKJavaScriptExceptionLineNumberErrorKey];
+    [userInfo setObject:RetainPtr { @(details->columnNumber) }.get() forKey:_WKJavaScriptExceptionColumnNumberErrorKey];
```
This seems like an error in the safer C++ checking. The result of `@()` should be usable without `RetainPtr` due either to autorelease or being an immortal object.

cdumez replied:
@rniwa See this feedback here and below. What should I do? Should I suppress and file radars? Or is it something the static analysis simply cannot determine is safe?

cdumez replied:
Since this is not hot code, I've been retaining to make static analysis happy but I agree with Darin I shouldn't have to.

rniwa replied:
oh, I already handle `@1` and `@"~"` but I'm not handling `@(~)` yet.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2268,7 +2268,7 @@ - (void)_didChangeEditorState
     id <WKUIDelegatePrivate> uiDelegate = (id <WKUIDelegatePrivate>)self.UIDelegate;
     if ([uiDelegate respondsToSelector:@selector(_webView:editorStateDidChange:)])
-        [uiDelegate _webView:self editorStateDidChange:dictionaryRepresentationForEditorState(_page->editorState())];
+        [uiDelegate _webView:self editorStateDidChange:RetainPtr { dictionaryRepresentationForEditorState(_page->editorState()) }.get()];
```
Shouldn’t `dictionaryRepresentationForEditorState` be returning a `RetainPtr`?

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2331,7 +2331,7 @@ - (WKNavigation *)loadFileRequest:(NSURLRequest *)request allowingReadAccessToUR
 - (void)setUnderPageBackgroundColor:(WebCore::CocoaColor *)underPageBackgroundColorOverride
 {
-    _page->setUnderPageBackgroundColorOverride(WebCore::roundAndClampToSRGBALossy(underPageBackgroundColorOverride.CGColor));
+    _page->setUnderPageBackgroundColorOverride(WebCore::roundAndClampToSRGBALossy(RetainPtr { underPageBackgroundColorOverride.CGColor }.get()));
```
This seems unfortunate, and like some kind of false positive. The `UIColor` `CGColor` property definitely returns an object that can be passed in to this function; the `RetainPtr` is not needed.

rniwa replied:
Hm... `CGColor` has `objc_returns_inner_pointer` not `ns_returns_autoreleased` so I don't think we can infer that from the method declaration.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3051,7 +3051,7 @@ - (void)_updateFixedContainerEdges:(const WebCore::FixedContainerEdges&)edges
 
     if (isTopColorChanging)
-        [self willChangeValueForKey:NSStringFromSelector(@selector(_sampledTopFixedPositionContentColor))];
+        [self willChangeValueForKey:RetainPtr { NSStringFromSelector(@selector(_sampledTopFixedPositionContentColor)) }.get()];
```
The result of `NSStringFromSelector` definitely does not need a `RetainPtr`; it’s too bad the validator doesn’t understand this.

rniwa replied:
Since there is no annotation on the function to indicate that, static analyzer can't figure out that. We could hard-code a bunch of similar functions and treat its return value as safe values though.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3069,7 +3069,7 @@ - (void)_updateFixedContainerEdges:(const WebCore::FixedContainerEdges&)edges
 
     if (isTopColorChanging)
-        [self didChangeValueForKey:NSStringFromSelector(@selector(_sampledTopFixedPositionContentColor))];
+        [self didChangeValueForKey:RetainPtr { NSStringFromSelector(@selector(_sampledTopFixedPositionContentColor)) }.get()];
```
The result of `NSStringFromSelector` definitely does not need a `RetainPtr`; it’s too bad the validator doesn’t understand this.

cdumez committed:
Address Safer CPP warnings in WKWebView.mm
https://bugs.webkit.org/show_bug.cgi?id=299688

Reviewed by Darin Adler.

* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(nsErrorFromExceptionDetails):
(-[WKWebView _initializeWithConfiguration:]):
(-[WKWebView _didInsertAttachment:withSource:]):
(-[WKWebView _didRemoveAttachment:]):
(-[WKWebView _didInvalidateDataForAttachment:]):
(-[WKWebView createWebArchiveDataWithCompletionHandler:]):
(dictionaryRepresentationForEditorState):
(-[WKWebView _didChangeEditorState]):
(-[WKWebView setUnderPageBackgroundColor:]):
(-[WKWebView _updateFixedContainerEdges:]):
(createUserInfo):
(-[WKWebView _dataTaskWithRequest:runAtForegroundPriority:completionHandler:]):
(-[WKWebView _loadAndDecodeImage:constrainedToSize:maximumBytesFromNetwork:completionHandler:]):
(-[WKWebView _decodeImageData:preferredSize:completionHandler:]):
(-[WKWebView _getMainResourceDataWithCompletionHandler:]):
(-[WKWebView _createWebArchiveForFrame:completionHandler:]):
(-[WKWebView _createWebArchiveForFrames:rootFrame:completionHandler:]):
(-[WKWebView _getApplicationManifestWithCompletionHandler:]):
(-[WKWebView _getTextFragmentMatchWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

Canonical link: https://commits.webkit.org/300671@main
